src/sources/soundsourceflac.cpp

```diff
@@ -345,6 +345,25 @@ FLAC__bool SoundSourceFLAC::flacEOF() {
     return m_file.atEnd();
 }
 
+inline
+FLAC__int32 adjustDecodedSample(FLAC__int32 decodedSample, SINT bitsPerSample) {
+    // Workaround for improperly encoded FLAC files that may contain
+    // garbage in the most significant, unused bits of decoded samples.
+    // Required at least for libFLAC 1.3.2. This workaround might become
+    // obsolete once libFLAC is taking care of these issues internally.
+    // https://bugs.launchpad.net/mixxx/+bug/1769717
+    // https://hydrogenaud.io/index.php/topic,61792.msg559045.html#msg559045
+    FLAC__int32 signBit = static_cast<FLAC__int32>(1) << (bitsPerSample - 1);
+    FLAC__int32 bitMask = (static_cast<FLAC__int32>(1) << bitsPerSample) - 1; // == (signBit << 1) - 1
+    FLAC__int32 maskedSample = decodedSample & bitMask;
+    if (maskedSample & signBit) {
```

Comment:
This is a relative hot loop.
```
decodedSample <<= 32 - bitsPerSample;
FloatSample = static_cast<CSAMPLE>(decodedSample) / 0xffffffff;
```
Should work fast, without any conditional code.

Reply:
:+1: I was already thinking if there is some bit magic to improve this workaround. First shifting into the other direction to move out the garbage and then catch the sign bit with a shift into the opposite direction came into my mind. The conditional is really ugly!

Reply:
One left shift + float multiplication is sufficient for the entire conversion to CSAMPLE! But I don't get your cast to floating point followed by the masking.